R/undesirable_function_linter.R

```diff
@@ -9,38 +9,49 @@
 undesirable_function_linter <- function(fun = default_undesirable_functions,
                                         symbol_is_undesirable = TRUE) {
   stopifnot(is.logical(symbol_is_undesirable))
+
   Linter(function(source_file) {
+    if (is.null(source_file$xml_parsed_content)) return(NULL)
     if (symbol_is_undesirable) {
-      vals <- c("SYMBOL_FUNCTION_CALL", "SYMBOL")
+      tokens <- c("SYMBOL_FUNCTION_CALL", "SYMBOL")
     } else {
-      vals <- "SYMBOL_FUNCTION_CALL"
+      tokens <- "SYMBOL_FUNCTION_CALL"
     }
-    lapply(
-      ids_with_token(source_file, vals, fun = `%in%`),
-      function(id) {
-        token <- with_id(source_file, id)
-        fun_name <- token[["text"]]
-        if (fun_name %in% names(fun)) {
-          line_num <- token[["line1"]]
-          start_col_num <- token[["col1"]]
-          end_col_num <- token[["col2"]]
-          msg <- sprintf("Function \"%s\" is undesirable.", fun_name)
-          alt_fun <- fun[[fun_name]]
-          if (!is.na(alt_fun)) {
-            msg <- c(msg, sprintf("As an alternative, %s.", alt_fun))
-          }
 
-          Lint(
-            filename = source_file[["filename"]],
-            line_number = line_num,
-            column_number = start_col_num,
-            type = "warning",
-            message = paste0(msg, collapse = " "),
-            line = source_file[["lines"]][[as.character(line_num)]],
-            ranges = list(c(start_col_num, end_col_num))
-          )
+    xpath <- paste0(
+      "//",
+      tokens,
+      "[",
+      paste("text()='", names(fun), "'", sep = "", collapse = " or "),
+      " and ",
+      "count(parent::expr/preceding-sibling::expr/SYMBOL_FUNCTION_CALL[text()='library' or text()='require'])=0",
+      "]",
```

Comment:
There seems to be a precedence error here.
Try linting `library(foo)` with c("foo", "bar") as undesired in that order to confirm.

Use parentheses around the `paste()` call to ensure correct precedence.